### PR TITLE
Implement connection and responde on server

### PR DESCRIPTION
**Pull request by AxelF44**
I've implemented the arrival of Client, Ai and GUI.

That's include of course the system of communication (write and read) with them.

### CODE REVIEW

File: server/include/server/client.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include "app/app.h"
+
+typedef struct s_client {
+    size_t fd;
+} client_t;
+
+/**
+ * @brief Create a client object.
+ *
+ * @param fd File descriptor of ai
```
Comment (mathieurobert1) [inline server/include/server/client.h:23]:
```suggestion
 * @param fd File descriptor of ai.
```

File: server/include/server/client.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * @return client_t* Struct client initialized
```
Comment (mathieurobert1) [inline server/include/server/client.h:24]:
```suggestion
 * @return client_t* Struct client initialized.
```

File: server/include/server/server.h
Comment (mathieurobert1) [inline server/include/server/server.h:1]: Be careful to some points at the end of doxygen sentences.

File: server/src/server/read/read_data.c [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    line[len_line] = current_char;
+    line[len_line + 1] = '\0';
+    return line;
+}
+
+char *read_line(int fd)
+{
+    char buffer[1024];
+    ssize_t bytes_received;
+    char *line = malloc(sizeof(char) * 1);
```
Comment (mathieurobert1) [inline server/src/server/read/read_data.c:29]: check your malloc

File: server/src/server/read/read_data.c [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+#include "ai/ai.h"
+#include "app/app.h"
+#include "server/client.h"
+
+char *append_char(char *line, char current_char)
+{
+    int len_line = 0;
+
+    len_line = strlen(line);
+    line = realloc(line, len_line + 2);
```
Comment (mathieurobert1) [inline server/src/server/read/read_data.c:19]: I'm not sure but you have to check realloc ?

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Overall that's a great work !

File: server/src/server/read/read_data.c (hunk 4 above)
Comment (Njord201) [reply server/src/server/read/read_data.c:19]: Check it

File: server/include/ai/ai.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
+** EPITECH PROJECT, 2024
+** Zappy Server
+** File description:
+** AI struct
```
Comment (Njord201) [inline server/include/ai/ai.h:5]: The file description is false.

File: server/include/ai/team.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    list_t *list_ai;
+    char *name;
+    size_t max_place;
+} team_t;
+
+/**
+ * @brief Create a team object.
+ *
+ * @param name Name of the team.
+ * @param max_place Max place in a team
```
Comment (Njord201) [inline server/include/ai/team.h:26]: Don't forget the point.


```suggestion
 * @param max_place Max place in a team.
```

File: server/include/ai/ai.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    list_t *list_messages;
+    size_t level;
+} ia_t;
+
+/**
+ * @brief Create a AI object.
+ *
+ * @param app Application with list of team.
+ * @param fd File descriptor of AI.
+ * @param team Team of AI
```
Comment (Njord201) [inline server/include/ai/ai.h:29]: Don't forget the point at the end of the sentence.


```suggestion
 * @param team Team of AI.
```

File: server/include/app/app.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
+/*
+** EPITECH PROJECT, 2024
+** Zappy Server
+** File description:
+** App struct
```
Comment (Njord201) [inline server/include/app/app.h:5]: The file description is false, not only the structure.

File: server/include/gui/gui.h
Comment (Njord201) [inline server/include/gui/gui.h:1]: Don't forget the point in the functions descriptions.

File: server/include/list/type.h
Comment (Njord201) [inline server/include/list/type.h:1]: Maybe rename the file in types.h

File: server/src/ai/ai.c [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    list_node_t *temp = app->teams_list->first;
+    team_t *team = NULL;
+    ia_t *ai = NULL;
+
+    while (temp) {
+        team = temp->data.team;
+        ai = check_ia(team, fd);
+        if (ai != NULL) {
+            return ai;
+        }
```
Comment (Njord201) [inline server/src/ai/ai.c:73]: Useless brackets

File: server/src/app/app.c [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+** app
+*/
+
+#include "app/app.h"
+
+static void create_teams(parsing_t *parsing, app_t *app)
+{
+    for (size_t index = 0; parsing->names[index]; index++) {
+        add_team(app->teams_list, parsing->names[index], parsing->clientsNb);
+    }
```
Comment (Njord201) [inline server/src/app/app.c:14]: Useless brackets

File: server/src/gui/gui.c [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+gui_t *find_gui(app_t *app, size_t fd)
+{
+    list_node_t *temp = app->gui_list->first;
+
+    while (temp) {
+        if (temp->data.gui->fd == fd) {
+            return temp->data.gui;
+        }
```
Comment (Njord201) [inline server/src/gui/gui.c:38]: Useless brackets

File: server/src/server/read/connection.c [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+    client_fd = accept(fd, (struct sockaddr *) &addr, &addr_len);
+    if (client_fd == -1)
+        return false;
+    data.client = create_client(client_fd);
+    if (data.client == NULL)
+        return false;
+    list_add_back(app->clients_list, data);
+    write(client_fd, "WELCOME\n", strlen("WELCOME\n"));
+    printf("NEW CONNECTION !\n");
```
Comment (Njord201) [inline server/src/server/read/connection.c:27]: Print in the stderr no ?

File: server/src/server/read/connection.c [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    node_data_t data;
+
+    client_fd = accept(fd, (struct sockaddr *) &addr, &addr_len);
+    if (client_fd == -1)
+        return false;
+    data.client = create_client(client_fd);
+    if (data.client == NULL)
+        return false;
+    list_add_back(app->clients_list, data);
+    write(client_fd, "WELCOME\n", strlen("WELCOME\n"));
```
Comment (Njord201) [inline server/src/server/read/connection.c:26]: Maybe put the string in a macro? One for the connection message and one for the length of the connection message?

File: server/src/server/read/read_data.c [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    line[len_line] = current_char;
+    line[len_line + 1] = '\0';
+    return line;
+}
+
+char *read_line(int fd)
+{
+    char buffer[1024];
+    ssize_t bytes_received;
```
Comment (Njord201) [inline server/src/server/read/read_data.c:28]: I don't think you need this variable (bytes_received), first of all you can only receive one byte in this configuration and you can put the read in if and while.

File: server/src/server/read/read_data.c [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        if (buffer[0] == '\n')
+            break;
+        line = append_char(line, buffer[0]);
+        memset(buffer, 0, sizeof(buffer));
+        bytes_received = read(fd, buffer, 1);
+    }
+    return line;
+}
+
+void handle_request(app_t *app, size_t fd, char *line)
```
Comment (Njord201) [inline server/src/server/read/read_data.c:46]: Why do you need the line variable?

File: server/src/server/read/read_data.c [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+{
```
Comment (Njord201) [inline server/src/server/read/read_data.c:47]: This function is weird is it finished?

File: server/src/server/read/read_data.c [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        server_quit_handler(app, fd);
+        printf("QUIT\n");
+        return true;
+    }
+    if (its_client(app, fd)) {
+        if (strcmp(line, "GRAPHIC\r") == 0) {
+            add_gui(app, fd);
+        } else {
+            add_ia(app, fd, line);
+        }
```
Comment (Njord201) [inline server/src/server/read/read_data.c:79]: Useless brackets

File: server/src/server/run.c [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+int handle_client_read(app_t *app, int fd)
+{
+    if (FD_ISSET(fd, &app->server->read_fds)) {
+        if (fd == app->server->fd) {
+            server_connection_handler(app, fd);
+        } else {
+            server_data_handler(app, fd);
+        }
```
Comment (Njord201) [inline server/src/server/run.c:67]: Useless

File: server/src/server/run.c [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+{
+    gui_t *gui = find_gui(app, fd);
+    ia_t *ai = find_ia(app, fd);
+
+    if (gui != NULL) {
+        write_message(gui->list_messages, gui->fd);
+    }
+    if (ai != NULL) {
+        write_message(ai->list_messages, ai->fd);
+    }
```
Comment (Njord201) [inline server/src/server/run.c:82]: Useless brackets

**Review by mathieurobert1 — APPROVED** [review]
Seems done :+1:

File: server/include/server/server.h
Comment (Njord201) [inline server/include/server/server.h:1]: Some points are missing with doxygen

File: server/src/ai/ai.c [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+    new_ia->list_messages = list_new();
+    nb_place = format_string("%ld\n", team->max_place -
+    (team->list_ai->len + 1));
+    map_size = format_string("%d %d\n", app->game->width, app->game->height);
+    add_message(new_ia->list_messages, nb_place);
+    add_message(new_ia->list_messages, map_size);
+    return new_ia;
+}
+
+bool add_ia(app_t *app, size_t fd, char *line)
```
Comment (Njord201) [inline server/src/ai/ai.c:30]: If you return true in all cases, why not define the function type as void?

File: server/src/app/app.c [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+** app
+*/
+
+#include "app/app.h"
+
+static void create_teams(parsing_t *parsing, app_t *app)
+{
+    for (size_t index = 0; parsing->names[index]; index++)
+        add_team(app->teams_list, parsing->names[index], parsing->clientsNb);
+
```
Comment (Njord201) [inline server/src/app/app.c:14]: Remove this useless line

File: server/src/gui/gui.c [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    gui_t *new_gui = malloc(sizeof(gui_t));
+
+    if (new_gui == NULL)
+        return NULL;
+    new_gui->fd = fd;
+    new_gui->list_messages = list_new();
+    return new_gui;
+}
+
+bool add_gui(app_t *app, size_t fd)
```
Comment (Njord201) [inline server/src/gui/gui.c:21]: Same here, you can change the return type of the function to void

File: server/src/server/run.c [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+        client = temp->data.client;
+        FD_SET(client->fd, &app->server->read_fds);
+        FD_SET(client->fd, &app->server->write_fds);
+        temp = temp->next;
+    }
+    reset_gui(app);
+    reset_ai(app);
+}
+
+int handle_client_read(app_t *app, int fd)
```
Comment (Njord201) [inline server/src/server/run.c:60]: You can put the return type of the function in void.

File: server/src/server/run.c [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+    if (FD_ISSET(fd, &app->server->read_fds)) {
+        if (fd == app->server->fd)
+            server_connection_handler(app, fd);
+        else
+            server_data_handler(app, fd);
+    }
+    return 0;
+}
+
+int handle_client_write(app_t *app, int fd)
```
Comment (Njord201) [inline server/src/server/run.c:71]: You can put the return type of the function in void.

**Review by Njord201 — CHANGES_REQUESTED** [review]
Some little request, good job !

**Review by Njord201 — APPROVED** [review]
Good job

**Review by mathieurobert1 — APPROVED** [review]

**Pull request merged**